[Michael Yang] iP

michaelyql opened the pull request:
# Terminator
> (Terminator): I will be back...

This is a chatbot inspired by the [Terminator](https://en.wikipedia.org/wiki/The_Terminator) franchise.

Features include
- Adding, deleting, marking of tasks

For example, `todo terminate aliens` creates a todo task for *terminating aliens* 🤖

In order to run the chatbot,
1. Download the latest jar release version into your preferred directory.
2. Run `java -jar Terminator` from the terminal or command line.

The entry point of the application is in the `main` function. 
```java
class Terminator {
    ...
    public static void main(String[] args) {
        Terminator tChatbot = new Terminator();
        tChatbot.run();
    }
}
```
<hr />

Additional features (to be added):
- [ ] Keep track of your ingredients in your kitchen fridge
- [x] Sweep your floors for you and keep your house clean

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

michaelyql committed:
Rename chatbot to Terminator and update greeting

michaelyql committed:
Add echo to chatbot

michaelyql committed:
Add list functionality

michaelyql committed:
add mark and unmark task functionality

michaelyql committed:
Add todo, event and deadline tasks

michaelyql committed:
Automate text UI testing

michaelyql committed:
Add custom exception handling

michaelyql committed:
Add delete functionality

michaelyql committed:
Merge branch 'add-gradle-support' into branch-Level-7

michaelyql committed:
Add save and Date/Time functionality

michaelyql committed:
Extract UI and Storage classes

michaelyql committed:
Organize project into packages

michaelyql committed:
Add JUnit tests

michaelyql committed:
Add JUnit tests for UI and MarkCommand

michaelyql committed:
Add javadoc comments

michaelyql committed:
Add search functionality

michaelyql committed:
Fix checkstyle errors

michaelyql committed:
Add basic JavaFX GUI

antgyf commented on src/main/java/terminator/Terminator.java:
```diff
@@ -0,0 +1,128 @@
+            try {
+                Command command = this.parser.parse(input);
+                String response = command.execute(this.taskList.getTaskList());
+                if (command instanceof TodoCommand
```
Would it be possible to implement this in the execute() method instead? ie. keep a static taskList field in the command class and call getTasksRemaining() on it instead. Not a big issue but something to consider.

michaelyql replied:
I feel that keeping a static `taskList` field in the `Command` class could lead to poorer separation of concerns, as the `Command` class is only intended to represent a command by the user, and hence it should not be concerned with how data is being stored.

antgyf commented on src/main/java/terminator/command/CommandParser.java:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * Creates a new instance of a CommandParser.
+     */
+    public CommandParser() {
```
Might be more efficient to keep these patterns as a static field and initialise on declaration.

antgyf commented on src/main/java/terminator/command/CommandParser.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public Command parse(String input) {
+        Matcher matcher;
+        if (input.isEmpty()) {
```
for many layers of if else statements it may be faster to use a switch statement!

michaelyql replied:
I decided to stick with `if/else` statements because `switch` statements require a target variable that is either a byte, short, char, int (and their wrappers), or String or enum. Switch statements can't match objects, so in this case it wouldn't make sense to try to match the `Matcher` object.

I feel that regex patterns are more robust and less error-prone in general for parsing user inputs, but I agree that it has performance downsides and is slower than regular switch statements.

antgyf commented on src/main/java/terminator/Terminator.java:
```diff
@@ -0,0 +1,128 @@
+    private void runTerminalUi() {
+        // Load data from storage
+        try {
+            this.storage.loadDataFromFile(this.taskList);
```
nit: can remove unnecessary use of this keyword. Should only use this keyword if there is ambiguity in code.

antgyf reviewed (commented):
Overall LGTM. Just a few small nits here and there!

xGladiate commented on src/main/java/terminator/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,51 @@
+     * @param todoList The task list.
+     * @throws TerminatorException if the description of the deadline task is empty.
+     */
+    @Override
```
When writing Javadocs for overridden methods, the `@inheritDoc` tag can be used to reuse the header comment from the parent method but with further modifications e.g., when the method has a slightly different behavior from the parent method. _Else, if f I am not wrong:_ The JavaDoc tool will use the comment from the overriding method.

xGladiate commented on src/main/java/terminator/task/Storage.java:
```diff
@@ -0,0 +1,246 @@
+                // Parse task description
+                String desc = data.substring(4);
+                Task task = null;
+                switch(taskType) {
+                case TODO:
+                    parseTodo(desc);
+                    break;
+                case EVENT:
+                    parseEvent(desc);
+                    break;
+                case DEADLINE:
+                    parseDeadline(desc);
+                    break;
+                default:
+                    // We shouldn't reach this condition since we already checked the TaskType beforehand
+                    break;
```
Good job on not making any common mistakes! 💯

xGladiate commented on src/main/java/terminator/Main.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @param stage The JavaFX Stage.
+     */
+    @Override
```
Good job grouping the code, which helps to provide clear segmentation of the different segments of code. 
**_Improvements_**: Might want to provide clear and concise explanations of what each section of code is doing. 

For example: 
``` java
    // Configure user input and send button
    userInput.setPrefWidth(325.0);
    sendButton.setPrefWidth(55.0);
```

xGladiate commented on src/main/java/terminator/Terminator.java:
```diff
@@ -0,0 +1,128 @@
+package terminator;
+
+import java.util.Scanner;
+
+import terminator.command.Command;
+import terminator.command.CommandParser;
+import terminator.command.DeadlineCommand;
+import terminator.command.DeleteCommand;
+import terminator.command.EventCommand;
+import terminator.command.TerminatorException;
+import terminator.command.TodoCommand;
+import terminator.task.Storage;
+import terminator.task.TaskList;
```
Good job for grouping the imports in a clear manner!

xGladiate reviewed (commented):
Overall, there is no visible violation of coding standards, keep up the good work! 💯

michaelyql committed:
Generate fat jar with GUI

michaelyql committed:
Refactor CLI and Controller into their own packages

CLI class and JavaFX controllers are in the main package.

This may lead to confusion and poor code organization.

Let's
* keep CLI only classes and JavaFX Controllers into their own packages

This will increase code modularity.

michaelyql committed:
Add Java Streams functionality

michaelyql committed:
Change parser command patterns to be final static

michaelyql committed:
Merge pull request #1 from michaelyql/branch-A-CodeQuality

Change parser command patterns to be final static

michaelyql committed:
Set up GitHub Actions workflow

Tests and checks are not being run when code is pushed to the
repository.

We should always run tests and checks to make sure that the application
works at each point of commit.

Let's
* use the workflow as defined in ./github/workflows/gradle.yml

This allows us to automate test and improve our efficiency.

michaelyql committed:
Add help command

michaelyql committed:
Add css styling and custom font

michaelyql committed:
Add UI screenshot

michaelyql committed:
Update UG and gh actions

michaelyql committed:
Add major GUI improvements

michaelyql committed:
Update docs

michaelyql committed:
Update main controller fxml file

michaelyql committed:
Fix data loading and storage

michaelyql committed:
Fix checkstyle and generate jar

michaelyql committed:
Add more error handling

michaelyql committed:
Add tests for event command
